activerecord/lib/active_record/attribute_assignment.rb

```diff
@@ -10,15 +10,18 @@ module AttributeAssignment
       def _assign_attributes(attributes)
         multi_parameter_attributes  = {}
         nested_parameter_attributes = {}
+        new_attributes = attributes.dup
 
-        attributes.each do |k, v|
-          if k.include?("(")
-            multi_parameter_attributes[k] = attributes.delete(k)
+        new_attributes.each do |k, v|
+          key = k.to_s
+
+          if key.include?("(")
+            multi_parameter_attributes[key] = new_attributes.delete(k)
           elsif v.is_a?(Hash)
-            nested_parameter_attributes[k] = attributes.delete(k)
+            nested_parameter_attributes[key] = new_attributes.delete(k)
           end
         end
-        super(attributes)
+        super(new_attributes)
 
         assign_nested_parameter_attributes(nested_parameter_attributes) unless nested_parameter_attributes.empty?
         assign_multiparameter_attributes(multi_parameter_attributes) unless multi_parameter_attributes.empty?
```

Comment:
Some other versions in no particular order.

Take 1:

```ruby
def _assign_attributes(attributes)
  multi_parameter_attributes  = {}
  nested_parameter_attributes = {}

  super \
    attributes.dup.delete_if do |key, value|
      key = key.to_s

      if key.include?("(")
        multi_parameter_attributes[key] = value
      elsif value.is_a?(Hash)
        nested_parameter_attributes[key] = value
      end
    end
  
  assign_nested_parameter_attributes(nested_parameter_attributes) unless nested_parameter_attributes.empty?
  assign_multiparameter_attributes(multi_parameter_attributes) unless multi_parameter_attributes.empty?
end
```

Take 2:

```ruby
def _assign_attributes(attributes)
  multi_parameter_attributes  = {}
  nested_parameter_attributes = {}
  new_attributes = {}

  attributes.each do |key, value|
    key       = key.to_s
    hash      = multi_parameter_attributes  if key.include?("(")
    hash    ||= nested_parameter_attributes if value.is_a?(Hash)
    hash    ||= new_attributes
    hash[key] = value
  end

  super(new_attributes)
  
  assign_nested_parameter_attributes(nested_parameter_attributes) unless nested_parameter_attributes.empty?
  assign_multiparameter_attributes(multi_parameter_attributes) unless multi_parameter_attributes.empty?
end
```

Take 1 taken a little further with the extra idea that nested or multi parameter attributes are passed rarely:

```ruby
def _assign_attributes(attributes)
  multi_parameter_attributes = nested_parameter_attributes = nil

  super \
    attributes.dup.delete_if do |key, value|
      key = key.to_s

      if key.include?("(")
        (multi_parameter_attributes ||= {})[key] = value
      elsif value.is_a?(Hash)
        (nested_parameter_attributes ||= {})[key] = value
      end
    end

  assign_nested_parameter_attributes(nested_parameter_attributes) if nested_parameter_attributes&.any?
  assign_multiparameter_attributes(multi_parameter_attributes) if multi_parameter_attributes&.any?
end
```

Reply:
Personally, I like take 1 better.